slo-monitor/src/monitors/pod_monitor.go

```diff
@@ -0,0 +1,356 @@
+	return &PodStartupLatencyDataMonitor{
+		kubeClient:     c,
+		PodStartupData: map[string]PodStartupMilestones{},
+	}
+}
+
+// Run starts a PodStartupLatencyDataMonitor: it creates all watches, populates PodStartupData and updates
+//  PodE2EStartupLatency metric
+func (pm *PodStartupLatencyDataMonitor) Run(stopCh chan struct{}) {
+	// When starting we need to get a 'current' ResourceVersion, so that
+	podsList, err := pm.kubeClient.Core().Pods(v1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		panic(err)
+	}
+	startPodRV, err := storage.ParseListResourceVersion(podsList.ResourceVersion)
+	if err != nil {
+		panic(err)
+	}
+	glog.Infof("Initial Pod RV: %v", startPodRV)
+	pm.startPodRV = startPodRV
+
+	eventList, err := pm.kubeClient.Core().Events(v1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		panic(err)
+	}
+	startEventRV, err := storage.ParseListResourceVersion(eventList.ResourceVersion)
+	if err != nil {
+		panic(err)
+	}
+	glog.Infof("Initial Event RV: %v", startEventRV)
+	pm.startEventRV = startEventRV
+
+	controller := NewWatcherWithHandler(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				obj, err := pm.kubeClient.Core().Pods(v1.NamespaceAll).List(options)
+				return runtime.Object(obj), err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return pm.kubeClient.Core().Pods(v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Pod{},
+		func(obj interface{}) error {
+			p, ok := obj.(*v1.Pod)
+			if !ok {
+				glog.Errorf("Failed to cast observed object to *v1.Pod: %v", obj)
+				return nil
+			}
+			pm.handlePodUpdate(p)
+			return nil
+		},
+		func(obj interface{}) error {
+			p, ok := obj.(*v1.Pod)
+			if !ok {
+				glog.Errorf("Failed to cast observed object to *v1.Pod: %v", obj)
+				return nil
+			}
+			pm.handlePodDelete(p)
+			return nil
+		},
+	)
+	go controller.Run(stopCh)
+
+	eventSelector := fields.Set{
+		"involvedObject.kind": "Pod",
+		// this is actually source.Component
+		"source": "kubelet",
+	}.AsSelector().String()
+
+	eventController := NewWatcherWithHandler(
+		&cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = eventSelector
+				obj, err := pm.kubeClient.Core().Events(v1.NamespaceAll).List(options)
+				return runtime.Object(obj), err
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = eventSelector
+				return pm.kubeClient.Core().Events(v1.NamespaceAll).Watch(options)
+			},
+		},
+		&v1.Event{},
+		func(obj interface{}) error {
+			e, ok := obj.(*v1.Event)
+			if !ok {
+				glog.Errorf("Failed to cast observed object to *v1.Event: %v", obj)
+				return nil
+			}
+			pm.handleEventUpdate(e)
+			return nil
+		},
+		func(obj interface{}) error {
+			return nil
+		},
+	)
+	go eventController.Run(stopCh)
+	go wait.Until(pm.purgeOutdated, 15*time.Second, stopCh)
+
+	select {
```

Comment:
Oh, yes - after I added purger I can remove this (it was needed, as I wanted this function to be blocking).